src/storage/on_disk.rs

```diff
@@ -0,0 +1,242 @@
+use std::{
+    path::{Path, PathBuf},
+    sync::Mutex,
+};
+
+use ahash::AHashMap;
+use alloy_primitives::{keccak256, Address, B256, B64, U256};
+use libmdbx::{
+    Database, DatabaseKind, DatabaseOptions, Mode, NoWriteMap, ReadWriteOptions, SyncMode,
+    TableFlags, WriteFlags,
+};
+
+use super::{AccountBasic, EvmAccount, EvmCode, Storage};
+
+/// A storage that reads data from an on-disk MDBX database.
+#[derive(Debug)]
+pub struct OnDiskStorage {
+    db: Database<NoWriteMap>,
+    cache_accounts: Mutex<AHashMap<Address, Option<EvmAccount>>>,
+    cache_bytecodes: Mutex<AHashMap<B256, Option<EvmCode>>>,
+    cache_block_hashes: Mutex<AHashMap<u64, B256>>,
+}
+
+impl OnDiskStorage {
+    /// Opens the on-disk storage at the specified path.
+    pub fn open(path: impl AsRef<Path>) -> Result<Self, libmdbx::Error> {
+        let db = Database::open_with_options(
+            &path,
+            DatabaseOptions {
+                max_tables: Some(16),
+                ..DatabaseOptions::default()
+            },
+        )?;
+        Ok(Self {
+            db,
+            cache_accounts: Mutex::default(),
+            cache_bytecodes: Mutex::default(),
+            cache_block_hashes: Mutex::default(),
+        })
+    }
+
+    fn load_account_and_then<T, F: Fn(&Option<EvmAccount>) -> Result<T, libmdbx::Error>>(
+        &self,
+        address: &Address,
+        op: F,
```

Comment:
Can we not just return an `EvmAccount` here?

Reply:
Because `EvmAccount` is heavy, and we cannot return `&EvmAccount`.

Reply:
I meant why do we go through `op` instead of just returning directly. And yes, returning a reference is good.

Reply:
I get this compile error when trying to return a `&EvmAccount`:

```
cannot return value referencing local variable `evm_account`
returns a value referencing data owned by the current
```

I am checking if dashmap's `Ref` give me any possibilities.

Reply:
Regardless of `DashMap` or `HashMap`, it should be possible by pushing `evm_account` into the cache then returning a reference into the cache.

Reply:
ok, now it's returning `OccupiedEntry<Address, Option<EvmAccount>>`